apps/web/src/server/app/auth/auth.router.js

```diff
@@ -0,0 +1,29 @@
+import express from 'express';
+import msalNode from '@azure/msal-node';
+import { authSignIn, authUnauthorized, authError, handleRedirect } from './auth.controller.js';
+
+const router = express.Router();
+const cryptoProvider = new msalNode.CryptoProvider();
+
+// The order here is important as the redirect needs to be handled before a new session nonce is set.
+// And a nonce should be set before anythig
+router.route('/redirect').get(handleRedirect);
+router.route('/redirect').post(handleRedirect);
+
+// TODO: This should actually be run for each main route, not just `/auth`.
+router.use((request, response, next) => {
+	if (!request.session) {
+		throw new Error('No session found for this request'); // TODO: Handle this gracefully
+	}
+
+	// Add session nonce for crsf
+	request.session.nonce = cryptoProvider.createNewGuid();
```

Comment:
How is this being used to prevent CSRF? Won't we have to start sending this from the UI with every post/put/patch?

Reply:
When you login there is a series of redirect going from our app to Azure, you can argue that when it comes back if its not the same session someone captured that redirect urls from Azure and they could authenticate and get the tokens from it.

I followed their best practices on how this can be handled. To be honest didn't read to much into the attack vectors examples for this particular thing but I imagine is just that as it makes sense from a possible issue.